### PR TITLE
refactor(core): avoid hydration warnings when `RenderMode.Client` is set

### PR DESCRIPTION
**Pull request by AndrewKushnir**
With the newly-added `RenderMode` config for routes, some of the routes may have the `RenderMode.Client` mode enabled, while also having `provideClientHydration()` function in provider list at bootstrap. As a result, there was a false-positive warning in a console, notifying developers about hydration misconfiguration.

This commit adds extra logic to handle this situation and avoid such warnings.

Note: there is a change required on the CLI side to add an extra marker, which would activate the logic added in this commit.

## PR Type
What kind of change does this PR introduce?

- [x] Refactoring (no functional changes, no api changes)

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/core/src/hydration/api.ts [hunk 1]
```diff
@@ -191,14 +192,30 @@ export function withDomHydration(): EnvironmentProviders {
...
+        if (!isPlatformBrowser()) {
+          return;
+        }
+
+        if (inject(IS_HYDRATION_DOM_REUSE_ENABLED)) {
+          // Since this function is used across both server and client,
+          // make sure that the runtime code is only added when invoked
+          // on the client. Moving forward, the `isPlatformBrowser` check should
+          // be replaced with a tree-shakable alternative (e.g. `isServer`
+          // flag).
```
Comment (JeanMeche) [inline packages/core/src/hydration/api.ts:204]: nit: we might want to move that comment to the previous `if` as we don't have the `isPlatformBrowser` here anymore.

File: packages/core/src/hydration/api.ts [hunk 2]
```diff
@@ -250,14 +267,16 @@ export function withI18nSupport(): Provider[] {
...
-      useValue: true,
+      useFactory: () => inject(IS_HYDRATION_DOM_REUSE_ENABLED),
     },
     {
       provide: ENVIRONMENT_INITIALIZER,
       useValue: () => {
-        enableI18nHydrationRuntimeSupport();
-        setIsI18nHydrationSupportEnabled(true);
-        performanceMarkFeature('NgI18nHydration');
+        if (inject(IS_HYDRATION_DOM_REUSE_ENABLED)) {
```
Comment (JeanMeche) [inline packages/core/src/hydration/api.ts:275]: Is this change (and the previous one) a separate/unrelated change ? 
(While being correct I agree, wondering if it should at least in its own commit).

**Review by JeanMeche — APPROVED** [review]
Beside the comments, LGTM !

File: packages/core/src/hydration/api.ts [hunk 3]
```diff
@@ -143,6 +142,21 @@ function whenStableWithTimeout(appRef: ApplicationRef, injector: Injector): Prom
...
+ * Defines a name of an attribute that is added to the <body> tag
+ * in the `index.html` file in case a given route was configured
+ * with `RenderMode.Client`. 'cm' is an abbreviation for "Client Mode".
+ */
+export const CLIENT_RENDER_MODE_FLAG = 'ngcm';
+
+/**
+ * Checks whether the `RenderMode.Client` was defined for the current route.
+ */
+function isClientRenderModeEnabled() {
```
Comment (alan-agius4) [inline packages/core/src/hydration/api.ts:155]: NIT: prefer using an explicit return type.

```suggestion
function isClientRenderModeEnabled(): boolean {
```

**Review by alan-agius4 — APPROVED** [review]
LGTM

File: packages/core/src/hydration/api.ts (hunk 2 above)
Comment (AndrewKushnir) [reply packages/core/src/hydration/api.ts:274]: The change was not need previously, since having `provideClientHydration()` for CSR apps was a misconfiguration. With the `RenderMode`, this becomes a supported feature, so all related code is updated accordingly.

**Comment by AndrewKushnir** [comment]
[Presubmit](http://test/OCL:680864239:BASE:680864269:1727761437611:fb501d2d).

**Comment by devversion** [comment]
This PR was merged into the repository by commit 222544491b32ffbdf301b11678f97d9e14afe323.

The changes were merged into the following branches: main
